### PR TITLE
Added `Itertools::dedup_with_count()` and `Itertools::dedup_by_with_count()`

### LINKED ISSUE #393: [RFC] `dedup()`/`dedup_by()` with count
Every once in a while I find myself re-implementing a "dedup with count", an iterator that instead of yielding `T` yields `(usize, T)`:

```rust
let data = vec![10, 10, 20, 30, 30, 30, 20];
assert_eq!(
    data.into_iter().dedup_with_count().collect(),
    vec![(2, 10), (1, 20), (3, 30), (1, 20)]
);
```

If you agree that these (`dedup_with_count()`/`dedup_by_with_count()`) are useful methods I can open a PR with the implementation soon.

### PR DESCRIPTION
**Pull request by orium**
Fixes #393.

### CODE REVIEW

File: src/adaptors/mod.rs [hunk 1]
```diff
@@ -803,6 +803,88 @@ impl<I, Pred> Iterator for DedupBy<I, Pred>
     }
 }
 
+/// An iterator adaptor that removes repeated duplicates, while keeping a count of how many
+/// repeated elements were present. This will determine equality using a comparison function.
+///
+/// See [`.dedup_by_with_count()`](../trait.Itertools.html#method.dedup_by_with_count) or
+/// [`.dedup_with_count()`](../trait.Itertools.html#method.dedup_with_count) for more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct DedupByWithCount<I, Pred>
```
Comment (orium) [inline src/adaptors/mod.rs]: Should this be on the `adaptors` module?

**Comment by orium** [comment]
@jswrenn Can you take a look when you have some time?

**Comment by jswrenn** [comment]
Sorry for the delay and for the contribution! My only nit is the name—it suggests that `count` will be the second component of the tuple. Perhaps `count_dedup` and `count_dedup_by`?

(This is yet another case where having rust-lang/rfcs#2584 would be helpful!)

**Comment by phimuemue** [comment]
Hi there! Should we try to implement this in terms of (a generalized) `Dedup`/`DedupBy`?

As far as I can see, it (naturally) shares quite a bit of implementation with `Dedup`, and we could possibly express these similarities in code. (E.g. why does `Dedup` use `CoalesceIter` internally, while `DedupWithCount` goes with `Peekable`? Could this lead to subtle differences in behavior?)

If we actually want implement it in terms of `Dedup`, one way I could imagine is to equip `DedupBy` with an additional customization point (read: type parameter) that can be used to do a side-computation in `next_with` resp. `fold`. In the plain old `Dedup`/`DedupBy`-case, these side-computations could just be no-ops, while for `DedupWithCount`/`DedupByWithCount` it could do the counting.

**Comment by orium** [comment]
@jswrenn 

> My only nit is the name—it suggests that count will be the second component of the tuple. Perhaps count_dedup and count_dedup_by?

I think it makes sense to start the function name with the fundamental thing that is going on, which is deduplication.  The count is the extra metadata.  I also like how an alphabetical order of the names of the functions naturally group these functions together (and also that autocompletion on `dedup` will list all these).

@phimuemue

Yeah, it makes sense to try to re-use code.  I've re-wrote it in terms of `CoalesceCore`, which I generalize slightly.  If I try to change `DedupBy` to accommodate a dedup with count it will either look like `CoalesceCore` or it feels a bit forced.  Also, the `fold` specialization for `DedupBy` doesn't fit very well with a dedup with count.

**Comment by phimuemue** [comment]
Kudos for going with `CoalesceCore`. Imho it looks better.

Imho, we should still strive to unify the implementations further, but your observation

> If I try to change `DedupBy` to accommodate a dedup with count it will either look like `CoalesceCore` or it feels a bit forced. Also, the `fold` specialization for `DedupBy` doesn't fit very well with a dedup with count.

made me think why we even have `CoalesceCore` as an extra struct in the first place. I think we could even try to unify the implementations of `Coalesce` and `Dedup`(`By`), simplifying the implementations quite a bit. This would essentially mean that `Dedup`(`By`) is just a special case of `Coalesce` (which - intuitively - seems reasonable to me), and (I think) `Dedup`(`By`)`WithCount` could then be specializations, too.

Doing so, however, may be a bit too much for this ticket.

**Comment by orium** [comment]
Yes, I think that's  perfectly possible.  I can look into it, but, as you suggest, I prefer to do that in a later PR.

**Comment by phimuemue** [comment]
> 
> 
> Yes, I think that's perfectly possible. I can look into it, but, as you suggest, I prefer to do that in a later PR.

I sketched an implementation for this in https://github.com/rust-itertools/itertools/pull/440. As stated there, I will rebase it if this PR is accepted.

File: src/adaptors/mod.rs (hunk 1 above)
Comment (jswrenn) [reply src/adaptors/mod.rs:812]: This location of `DedupByWithCount` is fine. The `adaptors/mod.rs` file is absolutely getting over-sized, though. A restructuring is overdue, but this PR doesn't need to do it.

**Comment by jswrenn** [comment]
Looks good to me! Thanks!

bors r+

**Pull request merged**